File: design-patterns/csharp/adapter/LegacyGatewayAdapter.cs
using System;

namespace Adapter
{
    public class LegacyGatewayAdapter : IPaymentProcessor
    {
        private readonly LegacyGateway legacyGateway;
        private bool isPaymentSuccessful;

        public LegacyGatewayAdapter()
        {
            legacyGateway = new LegacyGateway();
        }

        public void ProcessPayment(double amount, string currency)
        {
            try
            {
                legacyGateway.ExecuteTransaction(amount, currency);
                isPaymentSuccessful = true;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error processing payment: {ex.Message}");
                isPaymentSuccessful = false;
            }
        }

        public bool IsPaymentSuccessful()
        {
            return isPaymentSuccessful && 
                   legacyGateway.CheckStatus(legacyGateway.GetReferenceNumber());
        }

        public string GetTransactionId()
        {
            return legacyGateway.GetReferenceNumber().ToString();
        }
    }
} 

File: design-patterns/cpp/adapter/legacy_gateway_adapter.cpp
#include "legacy_gateway_adapter.h"
#include <iostream>

LegacyGatewayAdapter::LegacyGatewayAdapter(LegacyGateway* legacyGateway) 
    : legacyGateway(legacyGateway) {}

void LegacyGatewayAdapter::processPayment(double amount, const std::string& currency) {
    std::cout << "LegacyGatewayAdapter: Processing payment of " << amount << " " << currency << std::endl;
    legacyGateway->executeTransaction(amount, currency);
    std::cout << "LegacyGatewayAdapter: Payment processed successfully. Txn ID: " 
              << legacyGateway->getReferenceNumber() << std::endl;
}

bool LegacyGatewayAdapter::isPaymentSuccessful() {
    return legacyGateway->checkStatus(legacyGateway->getReferenceNumber());
}

std::string LegacyGatewayAdapter::getTransactionId() {
    return std::to_string(legacyGateway->getReferenceNumber());
} 